pkg/referrerstore/oras/oras.go

```diff
@@ -363,63 +352,78 @@ func (store *orasStore) GetReferenceManifest(ctx context.Context, subjectReferen
 	if err != nil {
 		return nil, err
 	}
 
-	desc, err = repository.Resolve(ctx, subjectReference.Original)
+	desc, err := repository.Resolve(ctx, subjectReference.Original)
 	if err != nil {
-		evictOnError(err, store, subjectReference.Original)
+		evictOnError(ctx, err, subjectReference.Original)
 		return nil, err
 	}
 
-	// add the subject descriptor to cache
-	store.subjectDescriptorCache.Store(desc.Digest, desc)
-	// add the repository client to the auth cache if all repository operations successful
-	store.addAuthCache(subjectReference.Original, repository, expiry)
-
 	return &ocispecs.SubjectDescriptor{Descriptor: desc}, nil
 }
 
 // evict from cache on non retry-able errors including 401 and 403
-func evictOnError(err error, store *orasStore, subjectReference string) {
+func evictOnError(ctx context.Context, err error, subjectReference string) {
+	cacheProvider := cache.GetCacheProvider()
+	// if cache provider is not enabled, return
+	if cacheProvider == nil {
+		return
+	}
 	var ec *errcode.ErrorResponse
 
 	if errors.As(err, &ec) && (ec.StatusCode == http.StatusForbidden || ec.StatusCode == http.StatusUnauthorized) {
-		store.authCache.Delete(subjectReference)
+		artifactRef, err := registry.ParseReference(subjectReference)
+		if err != nil {
+			logrus.Warnf("failed to evict credential from cache for %s: %v", subjectReference, err)
+		}
+		cacheProvider.Delete(ctx, fmt.Sprintf(cache.CacheKeyOrasAuth, artifactRef.Registry))
 	}
 }
-func createDefaultRepository(ctx context.Context, store *orasStore, targetRef common.Reference) (registry.Repository, time.Time, error) {
-	if store.authProvider == nil || !store.authProvider.Enabled(ctx) {
-		return nil, time.Now(), fmt.Errorf("auth provider not properly enabled")
-	}
 
-	if entry, ok := store.authCache.Load(targetRef.Original); ok {
-		// if the auth cache entry expiration has not expired or it was never set
-		cacheEntry := entry.(authCacheEntry)
-		if cacheEntry.expiresOn.IsZero() || cacheEntry.expiresOn.After(time.Now()) {
-			return cacheEntry.client, cacheEntry.expiresOn, nil
-		}
+func createDefaultRepository(ctx context.Context, store *orasStore, targetRef common.Reference) (registry.Repository, error) {
+	if store.authProvider == nil || !store.authProvider.Enabled(ctx) {
+		return nil, fmt.Errorf("auth provider not properly enabled")
 	}
-
-	authConfig, err := store.authProvider.Provide(ctx, targetRef.Original)
+	artifactRef, err := registry.ParseReference(targetRef.Original)
 	if err != nil {
-		logrus.Warningf("auth provider failed with err, %v", err)
-		logrus.Info("attempting to use anonymous credentials")
+		return nil, err
+	}
+	var authConfig authprovider.AuthConfig
+	cacheProvider := cache.GetCacheProvider()
+	var cacheResponse string
+	found := false
+	if cacheProvider != nil {
+		cacheResponse, found = cacheProvider.Get(ctx, fmt.Sprintf(cache.CacheKeyOrasAuth, artifactRef.Registry))
+	}
+	if cacheResponse != "" && found {
```

Comment:
I feel line 402-420 could be under `cacheProvider != nil` case. If it's nil, we won't do the following logics, is that correct?

Reply:
It's slightly different. There are 2 cases that result in cache miss. 1. cacheProvider is not initialized 2. Cache does not have specified key. For both of these cases, we want to perform the same fallback